### ISSUE TITLE
H.264 audio: detect whether hardware is audio-capable

### ISSUE THREAD

**Issue by jotaen4tinypilot** [tiny-pilot/tinypilot#1276]
We are about to make audio available to all users that have audio-capable hardware. Currently, that’s the Voyager 2a devices, whose tc358743-chip support audio streaming.

For controlling audio via the app UI, we need to offer controls and indicators. These, however, should ideally only ever be present on devices with audio support. Otherwise, it’s distracting or even confusing for users to see “dead” or useless controls.

It’s not clear yet how we can detect audio-capability at runtime. The purpose of this ticket is to figure this out.

---

Note: Originally, this ticket was the UI proposal. However, we identified the hardware detection issue as blocking issue, so the following discussion was mainly about that. Therefore, I repurposed this ticket to make it about the hardware detection, and extracted the UI proposal to https://github.com/tiny-pilot/tinypilot/issues/1277.

**Comment by mtlynch** [comment]
>We could distinguish 3 cases:
>
>1.    Audio is generally available and on (unmuted)
>1.    Audio is generally available, but off (muted)
>1.    Audio is not available (e.g. in MJPEG mode, or in H.264 but without the appropriate hardware)

Can we actually identify case (3) in H.264 mode? I thought that even if the hardware doesn't support it, uStreamer will still generate an audio stream, but it won't have any data to feed it.

**Comment by mtlynch** [comment]
Can we spend an hour or two investigating whether it's possible to detect whether the audio stream contains anything? Or if the hardware supports audio at all?

If we start showing an "Unmute" button to users without audio-enabled hardware, they're going to be disappointed and confused when nothing happens. We might have to do that, but I'd like to avoid it if we can.

**Comment by mtlynch** [comment]
>Since we are in control of the hardware that we ship, we know which devices support audio and which doesn’t. So would it be viable for us to provide an audio flag somewhere? Like, in the simplest case, ship the audio-ready hardware with an [app setting](https://github.com/tiny-pilot/tinypilot/blob/master/dev_app_settings.cfg) that enables the audio-related UI (which otherwise would be hidden by default)?

That's definitely possible, but it's kind of the option of last resort because it means that we have to start building three TinyPilot Pro images, and then when users download the image, we have to split it into "Hobbyist" / "Voyager 1 and 2" / "Voyager 2a and later".

One last idea: can you try the latest TinyPilot Pro dev microSD image on both the old Voyager 2 and the audio-enabled Voyager 2 and compare these two things to look for markers of the new hardware:

* `sudo v4l2-ctl --all`
* Contents of `cat /sys/class/video4linux/video0` directory tree
  * Not sure if you can actually read all of these or if some of them are data streams

**Comment by mtlynch** [comment]
I've also asked our hardware engineering partner (who did the initial audio proof of concept) to see if there's a way to detect in software whether the audio breakout board is connected to the C790 chip.

**Comment by mtlynch** [comment]
>That’s inverse to what I naively expected. The bad news is, the value on Voyager 2a hardware flips to 1 as soon as you start to stream an audio signal. On Voyager 2, it stays 1 the entire time. I don’t fully understand what this is about yet, but my feeling is that it might not be safe to rely on this.

Yeah, I was thinking we could do something like check at TinyPilot app startup (or maybe in some systemd startup script after loading the EDID?) and write a file like `/home/tinypilot/audio-enabled` if it looks like a 2a, but that also feels super brittle.

>It’s probably worth to wait first, what the hardware people say.

Yeah, they're saying they can investigate this week, so hopefully we have an answer soon.

>Shall we create a spin-off ticket for the hardware detection issue? I feel that the basic UI is mostly independent.

Sure, that's a good idea. Can you do that?

The UI is blocked on the detection question for me, though, because if we can't reliably distinguish between audio-capable and audio-incapable hardware, then I definitely don't want to put something in the main UI that's meaningless/misleading to everyone without audio-ready hardware.

**Comment by mtlynch** [comment]
Closing as won't fix since it doesn't seem practical.

**Comment by mtlynch** [comment]
>We then also need to abandon #1277, right?

Ah, right. Correct. We can potentially resurrect it if users ask for ways to mute audio and we're not satisfied with burying it in a settings dialog.
